examples/pytorch/GATNE-T/src/main.py

```diff
@@ -0,0 +1,325 @@
+        self.trans_weights_s1 = Parameter(torch.FloatTensor(edge_type_count, embedding_u_size, dim_a))
+        self.trans_weights_s2 = Parameter(torch.FloatTensor(edge_type_count, dim_a, 1))
+
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        self.node_embeddings.data.uniform_(-1.0, 1.0)
+        self.node_type_embeddings.data.uniform_(-1.0, 1.0)
+        self.trans_weights.data.normal_(std=1.0 / math.sqrt(self.embedding_size))
+        self.trans_weights_s1.data.normal_(std=1.0 / math.sqrt(self.embedding_size))
+        self.trans_weights_s2.data.normal_(std=1.0 / math.sqrt(self.embedding_size))
+
+    # embs: [batch_size, embedding_size]
+    def forward(self, block):
+        input_nodes = block.srcdata[dgl.NID]
+        output_nodes = block.dstdata[dgl.NID]
+        batch_size = block.number_of_dst_nodes('user')
+        node_embed = self.node_embeddings
+        node_type_embed = []
+
+        with block.local_scope():
+            for i in range(self.edge_type_count):
+                edge_type = self.edge_types[i]
+                block.srcnodes['user'].data[edge_type] = self.node_type_embeddings[input_nodes, i]
+                block.dstnodes['user'].data[edge_type] = self.node_type_embeddings[output_nodes, i]
+                block.update_all(fn.copy_u(edge_type, 'm'), fn.sum('m', edge_type), etype=edge_type)
+                node_type_embed.append(block.dstnodes['user'].data[edge_type])
+        
+            node_type_embed = torch.stack(node_type_embed, 1)  
+            tmp_node_type_embed = node_type_embed.unsqueeze(2).view(-1, 1, self.embedding_u_size)  
+            trans_w = self.trans_weights.unsqueeze(0).repeat(batch_size, 1, 1, 1).view(
+                -1, self.embedding_u_size, self.embedding_size
+            )
+            trans_w_s1 = self.trans_weights_s1.unsqueeze(0).repeat(batch_size, 1, 1, 1).view(
+                -1, self.embedding_u_size, self.dim_a
+            )
+            trans_w_s2 = self.trans_weights_s2.unsqueeze(0).repeat(batch_size, 1, 1, 1).view(-1, self.dim_a, 1)
+
+            attention = F.softmax(
+                torch.matmul(
+                    torch.tanh(torch.matmul(tmp_node_type_embed, trans_w_s1)), trans_w_s2
+                ).squeeze(2).view(-1, self.edge_type_count),
+                dim=1,
+            ).unsqueeze(1).repeat(1, self.edge_type_count, 1)  
+            
+            node_type_embed = torch.matmul(attention, node_type_embed).view(-1, 1, self.embedding_u_size) 
+            node_embed = node_embed[output_nodes].unsqueeze(1).repeat(1, self.edge_type_count, 1) + \
+                torch.matmul(node_type_embed, trans_w).view(-1, self.edge_type_count, self.embedding_size)
+            last_node_embed = F.normalize(node_embed, dim=2)
+        
+            return last_node_embed  # [batch_size, edge_type_count, embedding_size]
+
+
+class NSLoss(nn.Module):
```

Comment:
what is this loss? please add comments. btw, please check out the previous comment. Python comments should be written under the class or function definition.